L74-java-channel-creds.md

```diff
@@ -0,0 +1,485 @@
+It could be a severe bug if certain TLS restrictions were ignored by a channel.
+To detect cases where the channel does not understand enough of the TLS
+configuration, there will be a method to check if there are required features
+that are not understood. If some are not understood, enough information will be
+returned to produce a useful error message.
+
+`TlsChannelCredentials` will support a no-arg constructor for the "all
+defaults" configuration. For any more advanced configuration, a builder would
+be used.
+
+```java
+package io.grpc;
+
+public final class TlsChannelCredentials extends ChannelCredentials {
+  /**
+   * Returns an empty set if this credential can be adequately understood via
+   * the features listed, otherwise returns a hint of features that are lacking
+   * to understand the configuration to be used for manual debugging.
+   *
+   * <p>An "understood" feature does not imply the caller is able to fully
+   * handle the feature. It simply means the caller understands the feature
+   * enough to use the appropriate APIs to read the configuration. The caller
+   * may support just a subset of a feature, in which case the caller would
+   * need to look at the configuration to determine if only the supported
+   * subset is used.
+   *
+   * <p>This method may not be as simple as a set difference. There may be
+   * multiple features that can independently satisfy a piece of configuration.
+   * If the configuration is incomprehensible, all such features would be
+   * returned, even though only one may be necessary.
+   *
+   * <p>An empty set does not imply that the credentials are fully understood.
+   * There may be optional configuration that can be ignored if not understood.
+   */
+  public Set<Feature> incomprehensible(EnumSet<Feature> understoodFeatures) {...}
+
+  /* Other methods can be added to support various Features */
+
+  public enum Feature {}
+
+  public final static class Builder {
+    public TlsChannelCredentials build() {...}
+  }
+}
+```
+
+An example `Feature` could be `CLIENT_CERTIFICATE`. When the `Feature` is
+added, methods `getCertificateChain()`, `getPrivateKey()`, `getPassword()` can
+be added. Observing the contents of those methods would be a requirement for
+`CLIENT_CERTIFICATE`. An implementation understanding `CLIENT_CERTIFICATE`
+might not support encrypted private keys and so could consider the credential
+unsupported if `getPassword()` returned non-`null`.
+
+To support very advanced use cases, Netty will provide a credential that wraps
+a `ProtocolNegotiator`. This allows implementations like ALTS and XDS to use
+internal APIs without forcing their users to use experimental or internal APIs,
+as their users would just interact with `ChannelCredentials`.
+
+```java
+package io.grpc.netty;
+
+final class NettyChannelCredentials extends ChannelCredentials {
+  public NettyChannelCredentials(ProtocolNegotiator.Factory negotiator) {...}
```

Comment:
Is it [`ProtocolNegotiatorFactory`](https://github.com/grpc/grpc-java/blob/master/netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java#L524)? I didn't find any `ProtocolNegotiator.Factory`.

Reply:
The Factory interface isn't all that interesting. Due to some builder holes I noticed that I needed a `getDefaultPort()` on the interface (basically it was a bug). It looks like it will be a new `ProtocolNegotiator.ClientFactory`. But it will be like ProtocolNegotiatorFactory in NettyChannelBuilder; basically the builder will be getting "out of the protocol negotiator business". I think we could actually remove the existing NettyChannelBuilder.ProtocolNegotiatorFactory fairly rapidly.